proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1108 @@
+unrestricted. However, wanted can, in principle, have different
+multiplicities.
+
+Let
+
+::
+
+   p :: Multiplicity
+   f :: A %p -> B
+   useC :: C %1 => A
+
+Then in
+
+::
+
+  f useC
+
+we have wanted ``C`` with multiplicity ``p`` (``p`` is a rigid
+variable). What do we do? We solve ``C`` as if it were an unrestricted
+wanted.
+
+Wanted with delayed multiplicities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Consider
+
+::
+
+   f :: A %p -> (A %p -> B) -> B
+   useC :: C %1 => A
+
+Then in
+
+::
+
+   f useC
+
+we have wanted ``C`` whose multiplicity is a unification variable, the
+value of which will be determined by the context. What do we do? There
+are two cases:
+- There's an unrestricted given with head ``C`` *and no such linear given*, then
+  the unrestricted given can solve the wanted.
+- There is a linear given with head ``C``: we don't solve ``C`` until
+  ``p`` has been determined.
+
+Quantified Constraints
+^^^^^^^^^^^^^^^^^^^^^^
+
+Our syntax extension naturally extends constraints in types' context
+to support linear implications ``C %1 => D`` when
+``-XQuantifiedConstraint`` is on. While this is not described in the
+paper_, our solving algorithm is based on `Efficient resource
+management for linear logic proof search`_, where such higher-order
+givens are handled. The extension is unproblematic, it would
+presumably be more effort to prevent it than to support it.
+
+Therefore, when ``-XLinearTypes`` and ``-XLinearConstraints`` are both
+on, contexts can contain implications of the form ``C %1 => D``.
+
+Existential types proposal
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The `Existential types proposal`_, if they ever materialise (🤞), will
+make this proposal even better. Using linear constraints in APIs most
+often require returning constraints too. For that we've been using
+GADTs. This was our simple mutable array API:
+
+::
+
+  read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a
+
+  free :: (Read n, Write n) %1 => MArray a n -> ()
+
+  data c /\ a where
+    Box :: c %1 => a -> c /\ a
+
+  new :: Linearly %1 => Int -> NewMArray a
+
+  data NewMArray a where
+    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n
+
+With this API we can write functions such as
+
+::
+
+  read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
+  read2AndDiscard arr =
+       let !(Box x)  = read arr 0
+           !(Box y)  = read arr 1
+           !()       = free arr
+       in (x, y)
+
+  linearly $
+    let
+      !(NewMArray arr1) = new
+      !(NewMArray arr2) = new
+    in
+    … -- modify the array as suited
+    Ur $ sum arr1 + sum arr2
+
+There is still a little bit of noise there, what with the defining of
+GADTs (while ``/\`` can be defined once and for all, types like
+``NewMArray`` must be defined for most every type because we lack
+type-level lambdas), and the constructors in the let-bindings. Plus
+this is cheating a tad: let bindings, even strict lets, don't
```

Comment:
As linear constraints have the same rules as linear functions, it is possible to try out such ideas using exising libraries and manual token passing.

WRT your particular example, current GHC always consider pattern guards non-linear consumption (https://play.haskell.org/saved/wvGeLDMJ)

Even with linear let it is not possible to consume linear value. This example works:
```haskell
read2AndDiscard :: Array a %1 -> Ur (a, a)
read2AndDiscard arr
  = let
  !(Ur x, arr') = read arr 0
  !(Ur y, arr'') = read arr' 
  in case consume arr'' of
    () -> Ur (x, y)
```
And this doesn't:
```haskell
read2AndDiscard :: Array a %1 -> Ur (a, a)
read2AndDiscard arr
  = let
  !(Ur x, arr') = read arr 0
  !(Ur y, arr'') = read arr' 
  !() = consume arr''
  in Ur (x, y)
```

Explicit bangs and consuming array in case do not work with pattern guards either.

Reply:
The second example is fixed in master. But since it was discovered after the 9.12 fork and is not a soundness bug, it'll only be possible to write it in 9.14.